feat: Add light and water commands

mpql opened the pull request:
# feat: Add light and water commands

## Description

Adds `light` and `water` commands symlinked to corresponding `light.sh` and `water.sh` bash scripts that act as abstraction layers to make on-demand control easier, and crontab scheduling shorter, simpler, and clearer.

Note that this is a draft while I monitor these changes in production on my own device.

Fixes #61 

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## How Has This Been Tested?

I'm unsure how to add meaningful and effective tests for the bash layer in the CI-CD step for this feature, but I'm certainly open to suggestions.

- [ ] ☹️?

## Checklist

- [x] My code follows the style guidelines of this project
  - mostly; next PR is actually going to be overhauling shell scripts to better comply with POSIX and pass `shellcheck`; my new code already does this 
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
  - Will do upon review, since these scripts are subject to change, pending review.
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
  - See above
- [x] New and existing unit tests pass locally with my changes

mpql commented:
It is worth noting that the `water` command takes an argument in minutes (`<1–15>`), turns on the pump, sleeps for the called duration, and automatically calls `water_off()` after the sleep returns. Considering this is meant to be an abstraction layer, this feels like a bit of a kludge to me, as I think this should happen in the pump controller (or possibly be a required argument unless called with e.g. `--run-continuously`), as I can't imagine it's desirable to run the water pump at 100% indefinitely.

It's also worth mentioning that the 15-minute limit can be tweaked, and instead of returning an error code on minutes > 15, I could simply respond by running it for 15 and generating e.g. a `Notice`-level log.

mpql commented:
It just occurred to me that I should get the location of the `light` and `water` scripts based on the location of the `setup.sh` -- I'll amend that later today.

iot-root commented:
1. I would avoid using environment HOME  , the setup uses 

```# Configuration
BIN_DIR=$(dirname $(readlink -f $0))
INSTALL_DIR=$(realpath $BIN_DIR/..)
```

2. for the water script, I suggest using a trap incase the script fails with pump running which would run the pump continuously.

```shell
# Kill background process and restore terminal when this shell exits
cleanup() {
    kill "$bgPid"
    stty "$original_settings"
    rm -f "$lock_file"
}
trap 'cleanup' EXIT
```

mpql commented:
Yeah, that's what I get for coding when I'm tired. And I'm showing I've learned nothing by repeating my mistake tonight / this morning. 😂

In any case, these should be a bit more robust. Still in draft while I test in prod locally, rough changelog:

- Corrected max brightness to 100%
- Tested and found min brightness (4%)
- Parameterized min / max / defaults
- Added options (default: `true`) for out-of-bounds inputs to run the pump / lights at defaults anyhow
- Getting bin path from installed script location
- Changed symlink path from `${HOME}/bin/` -> `${HOME}/.local/bin/` since raspi uses a `systemd` layout
- Usage text refers to variables for defaults and extrema
- Errors are more clear, and tell you what was wrong about the input, and reminds you of extrema
- More and better code comments
- `set -euo pipefail` to catch all errors (which I usually do; forgot first time around)
- Trapped various signals and pseudo-exit (which I'd also usually do) for `water` to ensure pump turns off in case of fatal errors
- Rewrote `water` parsing to accept times as e.g. 5 / 5s / 5m / 1m5s, which all gets converted to seconds at the end for clarity
- Changed formatting variables: `IT` -> `ITL` ; `NC` -> `RST` to better match pre-existing conventions

Currently using this crontab:

```sh
# Set lights to 35% at 6am, 70% at 7am, 35% at 9pm, and turn off at 10pm
0 6,21          * * *           "${HOME}/.local/bin/light" 35
0 7             * * *           "${HOME}/.local/bin/light" 70
0 22            * * *           "${HOME}/.local/bin/light" off

# Run water pump at 8am, 4pm, and 12am for 5 minutes
0 0,8,16        * * *           "${HOME}/.local/bin/water" 5m

# Collect sensor data every 30 mins
*/30 *          * * *           "${HOME}/garden-of-eden/bin/get-sensor-data.sh"
```

Hope this checks boxes, but if you've got any feedback, I'm all ears! Will mark ready for review if gardyn crons are successful!

mpql commented:
`cron` doesn't like the water script for some reason (it runs fine in shell), will debug when I'm more awake

mpql commented:
Never mind, all works fine, just forgot to update the scripts locally. 😂

Tossing around some ideas for testing and mocking to meet ci-cd requirements, but the scripts work fine at present.

iot-root commented on bin/setup.sh:
```diff
@@ -219,6 +219,14 @@ function check_i2c_sensors {
+    if [[ ! -d "${HOME}/.local/bin/" ]] ; then
+        mkdir "${HOME}/.local/bin/" ;
+    fi
+    ln -fs "${INSTALL_DIR}/bin/light.sh" "${HOME}/.local/bin/light"
```
the home local bin directory doesnt exist in $PATH environment variable by default. Lets use /usr/local/bin
ocho $PATH
/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/local/games:/usr/games

mpql replied:
It's conditional on the directory existing, creating the directory and `source`ing would indeed add it to the path. That said, `/usr/local/bin/` is also fine.

iot-root commented:
@mpql so I was unable to push because you forked, but I included your changes with some modifications

iot-root reviewed (commented):
simplified water script to just accept seconds eliminating input validation in shell

iot-root commented:
> @mpql so I was unable to push because you forked, but I included your changes with some modifications

nevermind, I used the browser merging to address the issue

mpql commented:
Ack, in the spirit of collaboration, I wish I'd had a chance to respond to some of this before you'd committed it! My current thoughts:

> simplified water script to just accept seconds eliminating input validation in shell

I was actually specifically intending to avoid seconds! 😅 The python files already accept seconds, and `water` was intended to be an abstraction layer for end user to both run directly without having to memorize the long commands, and to make things like crontab + their own scripts easier. In terms of use case, folks generally are going to be watering in minutes, though I had coded that to accept and interpret inputs like e.g. `1m30s`, which `sleep` accepts natively.

With `light`, have you tested below 4%? On my Gardyn 3.0, the lights flash on and right back off.

Also with `light`, I was also assuming Gardyn had a reason for not allowing the lights to go above 70%, either in terms of plant tolerance or life expectancy of the lights themselves, and figured if they wanted to exceed that, they could call the python directly instead of the friendly script.

I'm fine with `readonly`, but ALL_CAPS variables are generally meant to belong to the environment, not the current script, so the lowercase was on purpose, as were the `subject_object` variable / `subject_verb` function naming conventions.

Tabs are better in bash because then the viewer can control indent size via whatever `EDITOR` they have set, with the exception of stdout, where you generally want two spaces `  ` for multi-line content for readability, and to indicate that it belongs to the above-run command, and isn't the output of something else, e.g. with `usage()`.

`case` is actually functionally identical to `if [ ...`, where `[` is an abstraction of `test`, whereas `[[` is a bash extension that is much more predictable, and again, was a purposeful choice. Anything that you know will run in bash (per our shebang) ought use `[[ ... ]]`.

The rest seems to be making bash more... pythonic? For some reason? `bash` is much closer to a procedural language than an object-oriented one, so you generally don't want to be declaring functions whose code will always run exactly once within the execution of the script.

Additionally, there's generally going to be some tinkerers who will change variables, so even though it's aesthetically pleasing to have all the variables at the top in one place, the general point is to keep the user-config ones at the top, though it's cool to mark that with e.g. `### CONFIG`, etc.

And one final point, `EXIT` is a `bash` pseudo-signal, and there are weird cases where it doesn't always fire. It's useful as a catch-all if you specifically know you're running in bash, but I had some other common signals in there to really ensure we turn the water off on failures, based on testing and killing it a whole bunch of ways (and having the two separate traps is part of that).


Anyhow, just wanted to make sure you understand my reasoning here. I will at least be using my own versions to preserve features on my gardyn, with a couple tweaks.
